docs/decisions/2022-07-test-fixtures.md

```diff
@@ -0,0 +1,90 @@
+# Prefer ephemeral test fixtures where possible
+
+* **Status:** approved
+* **Last Updated:** 2022-07
+* **Objective:** Clarify the use of permanent vs. reusable Cloud resources
+
+## Context & Problem Statement
+
+Emblem evolved to use long-lived, singleton test fixtures in end-to-end (E2E) test runs. These include resources such as Google Cloud projects, Cloud Storage buckets, container images in Artifact Registry, and Cloud Run services.
+
+We decided to investigate whether we should migrate these long-lived, singleton fixtures to a more ephemeral approach.
+
+## Priorities
+
+Ideally, our test fixtures would be designed such that:
+
+* Test infrastructure is quick to set up.
+* Test infrastructure is documented (e.g. via Terraform files).
+* Tests themselves are quick to write.
+* Both tests and their infrastructure require minimal maintenance.
+* Tests are runnable by multiple [teams of] developers in parallel.
+* Tests cover as much of the application surface as possible.
+* Tests allow developers and contributors to quickly identify root causes of bugs.
+* Tests have minimal reliance on existing state.
+
+Of course, some of these goals are mutually exclusive: supporting parallel test runs often increases the cost of writing (and in some cases maintaining) tests and test infrastructure.
+
+## Constraints
+
+### Some resources MUST be singletons
+While most of Emblem's Google Cloud resources are provisioned programmatically using Terraform, some (such as the Google Cloud projects themselves) are currently intended to be provisioned manually.
+
+> Consequently, we can't rely entirely on automated provisioning systems (at least, not without significant further engineering and/or documentation work) to spin the entire project up and down arbitrarily.
+
+### Emblem is owned by a single team
+While Emblem is designed with a service-oriented architecture, all of Emblem's components are (currently) owned by the same team.
+
+> Consequently, optimizing our test fixtures for multi-team collaboration (such as large numbers of parallel test runs) isn't as important as minimizing test development/maintenance costs.
+
+## Considered Options
+
+In this case, we have a _continuous spectrum_ of options. Thus, the "options" below are more _extremes_ than options:
+
+* **Extreme 1:** All fixtures are centralized, long-lived singletons. Under no circumstances should 2+ instances of the same fixture exist.
+* **Extreme 2:** All fixtures are ephemeral and short-lived, and pluralistic (multiple instances can simultaneously coexist). 
+
+## Decision
+
+We'll use _programmatically provisioned_ fixture resources wherever practical. These resources can be made ephemeral and/or pluralistic if necessary.
+
+Examples include:
```

Comment:
Where do we have these examples? We're managing most of these in terraform.

Reply:
Some of these are removed through the [cleanup script](https://github.com/GoogleCloudPlatform/emblem/blob/main/installation-testing/terraform/cleanup.sh).

(I added a link to it here, though.)